src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -331,6 +334,74 @@ struct ScalarOp<ONNXTanOp> {
   using IOp = NotSuportedScalarOp;
 };
 
+//===----------------------------------------------------------------------===//
+// Scalar unary ops for lowering ONNXGeluOp
+//===----------------------------------------------------------------------===//
+template <>
+struct ScalarOp<ONNXGeluOp> {
+  using FOp = CustomScalarOp;
+  using IOp = CustomScalarOp;
+};
+
+template <>
+double analyzeSimdFor<ONNXGeluOp>(
+    Type t, Operation *op, int64_t &von, int64_t &son) {
+  double results;
+  StringRef approximate = dyn_cast<ONNXGeluOp>(op).getApproximate();
+  if (approximate == "none")
+    results = simdAnalysis({GenericOps::ArithmeticGop, GenericOps::DivGop,
+                               GenericOps::MulGop, GenericOps::SqrtGop},
+        {1, 1, 2, 1}, t, von, son);
+  if (approximate == "tanh")
+    results = simdAnalysis(
+        {GenericOps::ArithmeticGop, GenericOps::DivGop, GenericOps::MulGop,
+            GenericOps::PowGop, GenericOps::SqrtGop},
+        {2, 1, 4, 1, 1}, t, von, son);
+  return results;
+}
+
+template <>
+Value emitScalarOpFor<ONNXGeluOp>(ConversionPatternRewriter &rewriter,
+    Location loc, Operation *op, Type elementType,
+    ArrayRef<Value> scalarOperands) {
+  Value operand = scalarOperands[0];
+
+  CheckIfCustomScalarOpIsSupported<ONNXGeluOp>(elementType);
+  MultiDialectBuilder<KrnlBuilder, MathBuilder> create(rewriter, loc);
+
+  // Approximate is a required attribute and should have a default value of
+  // "none". "approximate = none" simply implies no approximation will take
+  // place. However, "approximate" can also have a string value of "tanh" which
+  // indicates the use of tanh approximation.
+  StringRef approximate = dyn_cast<ONNXGeluOp>(op).getApproximate();
+
+  // Create constants
+  Value half = create.math.constant(elementType, 0.5);
+  Value one = create.math.constant(elementType, 1);
+  Value three = create.math.constant(elementType, 3);
+  Value decimal = create.math.constant(elementType, 0.044715);
+  Value sqrt_two_over_pi = create.math.constant(elementType, sqrt(2 / M_PI));
+  Value sqrt_two = create.math.constant(elementType, sqrt(2));
+
+  // Calculations
+  Value div = create.math.div(operand, sqrt_two);
+  Value erfApprox = create.math.erf(div);
+  Value add = create.math.add(one, erfApprox);
+  Value dec = create.math.add(
+      operand, create.math.mul(decimal, create.math.pow(operand, three)));
+  Value tanhApprox = create.math.tanh(create.math.mul(sqrt_two_over_pi, dec));
+  // Approximate = none returns an output of y = 0.5 * x * (1 +
+  // erf(x/sqrt(2)))
+  if (approximate == "none")
+    return create.math.mul(create.math.mul(half, operand), add);
```

Comment:
To avoid emitting redundant code, please move local variables for the case `approximate == "none"` into the body of this `if` since they are not used in the other `if (approximate == "tanh")`, e.g. the following local variables.
```
  Value sqrt_two = create.math.constant(elementType, sqrt(2));
  Value div = create.math.div(operand, sqrt_two);
  Value erfApprox = create.math.erf(div);
  Value add = create.math.add(one, erfApprox);
```

Reply:
None of these values are redundant.... I actually only use each one once. If I move it into the body I get the error as shown below :)

Reply:
@hamptonm1, the code as is is a bit difficult to read. What @tungld is requesting is that the values that are exclusively computed for the first case be only included in the first case, and same for the second case. 

I have a hard time verifying the cost model because not all the computation before the if are used in both cases. 

If too difficult to separate, then I would suggest to simply have all of the code in the then, and then all of the code in the else, and removing values that are unused.

Reply:
Ohh this is not to be dogmatic, I was saying I received an error when attempting to do things the way @Tung suggested. I will try to find a better solution... I am open at this point. I understand his request and agree but I was stating the values are not redundant just in case. Thanks!

Reply:
Okay I figured out why it was failing.... it needed curly braces around the if statement or else we will get those errors. Everything is fixed now.